### PR TITLE
Use System.CommandLine for CLI arguments

### PR DESCRIPTION
**Pull request by Antonstockmarr**
Arguments are interpreted by the ArgumentHandler using the framework System.CommandLine. All options are unchanged. This makes maintenance and extension easier.

### CODE REVIEW

File: src/CLI/Utilities/ArgumentHandler.cs [hunk 1]
```diff
@@ -9,12 +10,88 @@ namespace CLI.Utilities
 {
     internal class ArgumentHandler
     {
-        internal async Task RunScripts(ArgumentParser argumentParser)
+        public RootCommand BuildRootCommand()
         {
+            var singleBoxDeployOption = new Option<bool>(
+            name: "--single-box-deploy",
+            getDefaultValue: () => false,
+            description: "Prepares and installs hub and spoke environments"
```
Comment (AndreasHassing) [inline src/CLI/Utilities/ArgumentHandler.cs:18]: [nit] move the arguments one indent in 😊

```suggestion
                name: "--single-box-deploy",
                getDefaultValue: () => false,
                description: "Prepares and installs hub and spoke environments"
```

Comment (AndreasHassing) [reply src/CLI/Utilities/ArgumentHandler.cs:18]: Also applies to below option definitions.

Comment (AndreasHassing) [inline src/CLI/Utilities/ArgumentHandler.cs:18]: Let's be a bit more descriptive here, I want the user to know that this deploys to the current machine and does not need additional VMs.
```suggestion
            description: "Prepares and installs hub and spoke environments on the current machine"
```

File: src/CLI/Utilities/ArgumentHandler.cs [hunk 2]
```diff
@@ -9,12 +10,88 @@ namespace CLI.Utilities
...
+            var hubDeployOption = new Option<bool>(
+            name: "--hub-deploy",
+            getDefaultValue: () => false,
+            description: "Prepares and installs the hub"
+            );
+
+            var spokeDeployOption = new Option<bool>(
+            name: "--spoke-deploy",
+            getDefaultValue: () => false,
+            description: "Prepares and installs the spoke"
```
Comment (AndreasHassing) [inline src/CLI/Utilities/ArgumentHandler.cs:30]:
```suggestion
            description: "Prepares and installs the spoke on the current machine"
```

File: src/CLI/Utilities/ArgumentHandler.cs [hunk 3]
```diff
@@ -9,12 +10,88 @@ namespace CLI.Utilities
...
+            var singleBoxDeployOption = new Option<bool>(
+            name: "--single-box-deploy",
+            getDefaultValue: () => false,
+            description: "Prepares and installs hub and spoke environments"
+            );
+            
+            var hubDeployOption = new Option<bool>(
+            name: "--hub-deploy",
+            getDefaultValue: () => false,
+            description: "Prepares and installs the hub"
```
Comment (AndreasHassing) [inline src/CLI/Utilities/ArgumentHandler.cs:24]:
```suggestion
            description: "Prepares and installs the hub on the current machine"
```

File: src/CLI/Utilities/ArgumentHandler.cs [hunk 4]
```diff
@@ -9,12 +10,88 @@ namespace CLI.Utilities
...
+
+            var spokeDeployOption = new Option<bool>(
+            name: "--spoke-deploy",
+            getDefaultValue: () => false,
+            description: "Prepares and installs the spoke"
+            );
+
+            var cleanStorageOption = new Option<bool>(
+            name: "--clean-storage",
+            getDefaultValue: () => false,
```
Comment (AndreasHassing) [inline src/CLI/Utilities/ArgumentHandler.cs:35]: Is the default, default not `false` already? Or?

**Review by AndreasHassing — CHANGES_REQUESTED** [review]
Minor changes requested, otherwise looks good.

One thing to consider: will Dragonfruit make the command line parsing more maintainable and readable?

We spend a lot of lines of code to create the options, put them in multiple method calls and run a risk to order them incorrectly. I think Dragonfruit fixes all of that?

**Comment by Antonstockmarr** [comment]
> Minor changes requested, otherwise looks good.
> 
> One thing to consider: will Dragonfruit make the command line parsing more maintainable and readable?
> 
> We spend a lot of lines of code to create the options, put them in multiple method calls and run a risk to order them incorrectly. I think Dragonfruit fixes all of that?

I tried working with Dragonfruit, it does remove a lot of the boilerplate of this solution. The problem was that I could not get it to produce proper help messages or descriptions. The --clean-storage option would have description cleanStorage and nothing more. But I think there might be a way to solve it, the readability and maintainability would be a lot better.

**Comment by AndreasHassing** [comment]
> > Minor changes requested, otherwise looks good.
> > One thing to consider: will Dragonfruit make the command line parsing more maintainable and readable?
> > We spend a lot of lines of code to create the options, put them in multiple method calls and run a risk to order them incorrectly. I think Dragonfruit fixes all of that?
> 
> I tried working with Dragonfruit, it does remove a lot of the boilerplate of this solution. The problem was that I could not get it to produce proper help messages or descriptions. The --clean-storage option would have description cleanStorage and nothing more. But I think there might be a way to solve it, the readability and maintainability would be a lot better.

Did you see this sample?
https://github.com/dotnet/command-line-api/blob/main/docs/Your-first-app-with-System-CommandLine-DragonFruit.md

If you add `/// <param ...` documentation, that should show up as documentation for each option.

**Comment by Antonstockmarr** [comment]
> > > Minor changes requested, otherwise looks good.
> > > One thing to consider: will Dragonfruit make the command line parsing more maintainable and readable?
> > > We spend a lot of lines of code to create the options, put them in multiple method calls and run a risk to order them incorrectly. I think Dragonfruit fixes all of that?
> > 
> > 
> > I tried working with Dragonfruit, it does remove a lot of the boilerplate of this solution. The problem was that I could not get it to produce proper help messages or descriptions. The --clean-storage option would have description cleanStorage and nothing more. But I think there might be a way to solve it, the readability and maintainability would be a lot better.
> 
> Did you see this sample? https://github.com/dotnet/command-line-api/blob/main/docs/Your-first-app-with-System-CommandLine-DragonFruit.md
> 
> If you add `/// <param ...` documentation, that should show up as documentation for each option.

Must have forgotten to compile or something, it works now.

File: src/CLI/CLI.csproj [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
     <PlatformTarget>x64</PlatformTarget>
+    <NoWarn>CS1591</NoWarn>
```
Comment (AndreasHassing) [inline src/CLI/CLI.csproj:12]: Why is this needed?

File: src/CLI/Program.cs [hunk 6]
```diff
@@ -1,29 +1,100 @@
...
+        /// <param name="spokeDeploy">Prepares and installs the spoke on the current machine</param>
+        /// <param name="cleanStorage">Cleans all Azure storage accounts for hub and spoke</param>
+        /// <param name="drainPipelines">Drains pipelines between hub and spoke</param>
+        /// <param name="startPipelines">Starts pipelines between hub and spoke</param>
+        public static async Task Main(bool singleBoxDeploy,
+                bool hubDeploy,
+                bool spokeDeploy,
+                bool cleanStorage,
+                bool drainPipelines,
+                bool startPipelines
```
Comment (AndreasHassing) [inline src/CLI/Program.cs]: In their example, they always set the default value, is there any benefit to that?

https://github.com/dotnet/command-line-api/blob/main/docs/Your-first-app-with-System-CommandLine-DragonFruit.md

File: src/CLI/Program.cs [hunk 7]
```diff
@@ -1,29 +1,100 @@
...
+        /// <summary>
+        /// Scale Unit Management DevTools
+        /// </summary>
+        /// <param name="singleBoxDeploy">Prepares and installs hub and spoke environments on the current machine</param>
+        /// <param name="hubDeploy">Prepares and installs the hub on the current machine</param>
+        /// <param name="spokeDeploy">Prepares and installs the spoke on the current machine</param>
+        /// <param name="cleanStorage">Cleans all Azure storage accounts for hub and spoke</param>
+        /// <param name="drainPipelines">Drains pipelines between hub and spoke</param>
+        /// <param name="startPipelines">Starts pipelines between hub and spoke</param>
+        public static async Task Main(bool singleBoxDeploy,
```
Comment (AndreasHassing) [inline src/CLI/Program.cs]:
```suggestion
        public static async Task Main(
            bool singleBoxDeploy,
```

**Review by AndreasHassing — APPROVED** [review]

File: src/CLI/CLI.csproj (hunk 5 above)
Comment (Antonstockmarr) [reply src/CLI/CLI.csproj:12]: It seems that dragonfruit sets a compiler flag about documentation to true, producing warnings for every class and method in the project that they do not have xml comments. Hence this suppression :). The warning was not enabled in the first place unless this doc flag is set.

File: src/CLI/Program.cs (hunk 6 above)
Comment (Antonstockmarr) [reply src/CLI/Program.cs]: The boolean options are false by default. Would make a lot of sense to set defaults for non-boolean options, but Im not sure it's necessary to set them all to false in this case?

Comment (AndreasHassing) [reply src/CLI/Program.cs]: Nah, that's probably fine.
